### PR TITLE
Upstream sync 2024 04 08

### PR DESCRIPTION
**Pull request by SageMoore**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -5,15 +5,18 @@ requires = [
     "ninja",
     "packaging",
     "setuptools >= 49.4.0",
-    "torch == 2.1.2",
+    "torch == 2.2.1",
```
Comment (mgoin) [inline pyproject.toml:8]: note that we will have to release a nightly version of magic-wand with this torch support to get testing working for this PR

File: Dockerfile [hunk 2]
```diff
@@ -79,17 +94,36 @@ WORKDIR /usr/src/flash-attention-v2
...
-FROM dev AS test
+# note that this uses vllm installed by `pip`
+FROM vllm-base AS test
 
-# copy pytorch extensions separately to avoid having to rebuild
-# when python code changes
-WORKDIR /vllm-workspace
-# ADD is used to preserve directory structure
 ADD . /vllm-workspace/
+<<<<<<< HEAD
```
Comment (SageMoore) [inline Dockerfile:126]: @mgoin @robertgshaw2-neuralmagic How much of this HEAD section do I need to keep

Comment (mgoin) [reply Dockerfile:126]: AFAIK the only thing added to this file is the 
```
# UPSTREAM SYNC: Install sparsity extras
RUN --mount=type=cache,target=/root/.cache/pip \
    pip install nm-magic-wand
```
section - so I would say accept everything from upstream and add that section

File: README.md [hunk 3]
```diff
@@ -6,8 +6,6 @@
 
 ## Installation
 
-The [nm-vllm PyPi package](https://pypi.org/project/nm-vllm/) includes pre-compiled binaries for CUDA (version 12.1) kernels, streamlining the setup process. For other PyTorch or CUDA versions, please compile the package from source.
```
[the remark this answers is not in the document]
Comment (SageMoore) [reply README.md:9]: Oh good catch. I'll add this back.

File: .github/actions/nm-build-vllm/action.yml [hunk 4]
```diff
@@ -33,7 +33,7 @@ runs:
       source $(pyenv root)/versions/${{ inputs.python }}/envs/${VENV}/bin/activate
       # TODO: adjust when we need a proper release. use nightly now.
       pip3 install --index-url http://${{ inputs.pypi }}:8080/ --trusted-host ${{ inputs.pypi }} nm-magic-wand-nightly
-      pip3 install -r requirements.txt
+      pip3 install -r requirements-cuda.txt
```
Comment (andy-neuma) [inline .github/actions/nm-build-vllm/action.yml:36]: wondering if this should be "requirements-build.txt"?

File: .github/scripts/build [hunk 5]
```diff
@@ -56,5 +56,5 @@ fi
 
 source $(pyenv root)/versions/${PYTHON}/envs/${VENV}/bin/activate
 pip3 install --index-url http://${PYPI_IP}:8080/ --trusted-host ${PYPI_IP} nm-magic-wand-nightly
-pip3 install -r requirements.txt
+pip3 install -r requirements-cuda.txt
```
Comment (andy-neuma) [inline .github/scripts/build:59]: same here, "requirements-build.txt"?

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -49,7 +49,7 @@ jobs:
       matrix:
           os: ['ubuntu-20.04']
           python-version: ['3.8', '3.9', '3.10', '3.11']
-          pytorch-version: ['2.1.2']  # Must be the most recent version that meets requirements.txt.
+          pytorch-version: ['2.2.1']  # Must be the most recent version that meets requirements-cuda.txt.
```
Comment (andy-neuma) [inline .github/workflows/publish.yml:52]: yeah

File: .github/workflows/scripts/build.sh [hunk 7]
```diff
@@ -9,7 +9,7 @@ LD_LIBRARY_PATH=${cuda_home}/lib64:$LD_LIBRARY_PATH
 
 # Install requirements
 $python_executable -m pip install wheel packaging
-$python_executable -m pip install -r requirements.txt
+$python_executable -m pip install -r requirements-cuda.txt
```
Comment (andy-neuma) [inline .github/workflows/scripts/build.sh:12]: "requirements-build.txt"?

File: Dockerfile [hunk 8]
```diff
@@ -38,18 +47,16 @@ RUN --mount=type=cache,target=/root/.cache/pip \
...
 
-# copy input files
+# files and directories related to build wheels
 COPY csrc csrc
 COPY setup.py setup.py
 COPY cmake cmake
 COPY CMakeLists.txt CMakeLists.txt
-COPY requirements.txt requirements.txt
+COPY requirements-common.txt requirements-common.txt
+COPY requirements-cuda.txt requirements-cuda.txt
```
Comment (andy-neuma) [inline Dockerfile:56]: do we need "requirements-dev.txt" or is this file being deprecated?

File: Dockerfile [hunk 9]
```diff
@@ -79,62 +94,62 @@ WORKDIR /usr/src/flash-attention-v2
...
-
-#################### TEST IMAGE ####################
-# image to run unit testing suite
-FROM dev AS test
-
-# copy pytorch extensions separately to avoid having to rebuild
-# when python code changes
+#################### vLLM installation IMAGE ####################
+# image with vLLM installed
+FROM nvidia/cuda:12.1.0-base-ubuntu22.04 AS vllm-base
```
Comment (andy-neuma) [inline Dockerfile:99]: yeah

**Review by andy-neuma — APPROVED** [review]
thanks

File: Dockerfile (hunk 8 above)
Comment (SageMoore) [reply Dockerfile:56]: Not sure. I'll check to see if it looks like we need anything in there.

**Pull request merged**